src/Dictionaries/HashedDictionary.cpp

```diff
@@ -35,16 +41,151 @@ namespace ErrorCodes
+{
+    using HashedDictionary = HashedDictionary<dictionary_key_type, sparse, sharded>;
+    using Queue = ConcurrentBoundedQueue<Block>;
+
+public:
+    explicit ParallelDictionaryLoader(HashedDictionary & dictionary_)
+        : dictionary(dictionary_)
+        , shards(dictionary.configuration.shards)
+        , simple_key(dictionary.dict_struct.getKeysSize() == 1)
+        , pool(shards)
+        , shards_queues(shards)
+    {
+        UInt64 backlog = dictionary.configuration.shard_load_queue_backlog;
+        LOG_TRACE(dictionary.log, "Will load the dictionary using {} threads (with {} backlog)", shards, backlog);
+
+        shards_slots.resize(shards);
+        std::generate(shards_slots.begin(), shards_slots.end(), [n = 0]() mutable { return n++; });
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            shards_queues[shard].emplace(backlog);
+            pool.scheduleOrThrowOnError([this, shard, thread_group = CurrentThread::getGroup()]
+            {
+                if (thread_group)
+                    CurrentThread::attachToIfDetached(thread_group);
+                setThreadName("HashedDictLoad");
+
+                threadWorker(shard);
+            });
+        }
+    }
+
+    void addBlock(Block block)
+    {
+        Blocks shards_blocks(shards);
+        for (size_t shard = 0; shard < shards; ++shard)
+            shards_blocks[shard] = block.cloneEmpty();
+
+        IColumn::Selector selector = createShardSelector(block, shards_slots);
+        splitBlock(selector, block, shards_blocks);
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            if (!shards_queues[shard]->push(std::move(shards_blocks[shard])))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to shards queue #{}", shard);
+        }
+    }
+
+    void finish()
+    {
+        for (auto & queue : shards_queues)
+            queue->finish();
+
+        Stopwatch watch;
+        pool.wait();
+        UInt64 elapsed_us = watch.elapsedMicroseconds();
+        LOG_TRACE(dictionary.log, "Processing the tail took {}ms", elapsed_us);
+    }
+
+    ~ParallelDictionaryLoader()
+    {
+        for (auto & queue : shards_queues)
+            queue->clearAndFinish();
+        pool.wait();
+    }
+
+private:
+    HashedDictionary & dictionary;
+    const size_t shards;
+    bool simple_key;
+    ThreadPool pool;
+    std::vector<std::optional<Queue>> shards_queues;
+    std::vector<UInt64> shards_slots;
+
+    void threadWorker(size_t shard)
+    {
+        Block block;
+        auto & shard_queue = *shards_queues[shard];
+
+        while (shard_queue.pop(block))
+        {
+            Stopwatch watch;
+            dictionary.blockToAttributes(block, shard);
+            UInt64 elapsed_us = watch.elapsedMicroseconds();
+            if (elapsed_us > 1'000'000)
+                LOG_TRACE(dictionary.log, "Block processing for shard #{} is slow {}ms (rows {}).", shard, elapsed_us, block.rows());
```

Comment:
In log `ms` is not microseconds, better to use milliseconds or rename to microseconds.